### ISSUE TITLE
Open MPI - Problem with retrieving values across name spaces

### ISSUE THREAD

**Issue by hppritcha** [openpmix/openpmix#1712]
A user at NERSC noticed that their app, which makes extensive use of ```MPI_Comm_spawn``` stopped working when switching from Open MPI 4.0.1 to 4.0.2/3.  The problem appears as an inability to exchange OMPI MODEX key/values across a namespace break, i.e. when the jobids of the opal_proc's differ.  The spawned processes (which have a different jobid for each set of spawned processes) cannot PMIx_Get key/values stored by the spawner.  git bisect indicates openpmix commit  2422fe85df4a98774e8c6e55a230bda991798f92 caused the application to stop working.  The spawned processes "know" about the jobid/vpid of the original spawner processes - they are using that in the MODEX_RECV call within OB1  BTLs.

The problem needs to be run at scale to show this behavior - 16 nodes or higher, although this may be an artifact of the way the app is using MPI_Comm_spawn and the way ORTE is scheduling the spawned processes across the SLURM allocation.

The problem is easily reproducible and consistently either fails or passes.

Note:
Open MPI 4.0.1 has embedded PMIx 3.1.2.
Open MPI 4.0.2 has embedded PMIx 3.1.4.
Open MPI 4.0.3 has embedded PMIx 3.1.5.

If I rebuild Open MPI 4.0.x against PMIx 3.1.3 the user application works again.  PMIx 3.1.3 does not have commit 2422fe85df4a98774e8c6e55a230bda991798f92.

Its not clear whether this is an issue in PMIx or something in the way Open MPI is handling spawn and the way it uses PMIx to support this feature.

NERSC ticket 152783

**Comment by rhc54** [comment]
FWIW: I tested the head of the v4.0.x branch on my local cluster (TCP only) and comm_spawn works just fine. I therefore expect it is either already fixed or this is something specific to the Cray.

**Comment by rhc54** [comment]
I _think_ I may have actually tracked this down. It appears that ORTE and PRRTE aren't actually doing the info exchange during `PMIx_Connect` as they should, which might be reflected in what PMIx was expecting in the various versions. I'm working on a fix for PRRTE - we'd have to see about backporting it to ORTE.

**Comment by rhc54** [comment]
Hmmm...no, I don't think this actually can be the root of the problem. Question for you: does their app work with OMPI master by any chance?

**Comment by rhc54** [comment]
Found it

**Comment by wckzhang** [comment]
I was able to reproduce this issue with mtt tests, specifically the ompi-tests/ibm/dynamic/ tests.

My mpirun command on the v4.1.x Open MPI branch:
mpirun --np 2 -N 1 --hostfile ~/hostfile --mca pml cm ~/ompi-tests/ibm/dynamic/spawn_argvnull_mpifh

The dynamic spawn tests have failed with internode runs (intranode works fine) since 3.1.4rc2.

I traced the errors to the vicinity of the commit referenced:
```
52b21e0e Fix a typo and add a little debug - fails                                                                                 
  Local host: ip-172-31-3-255                                                   
  Remote host: (null)                                                           
  Error: Not found (-13)
2422fe85 Resolve get of proc-specific job-level info from another nspace - PMIX ERROR: NOT-FOUND in file client/pmix_client_get.c at line 334
7c6cd3a5 Merge pull request #1371 from rhc54/cmr31/upd - PMIX ERROR: NOT-FOUND in file client/pmix_client_get.c at line 334
c08fb041 Update NEWS - PMIX ERROR: NOT-FOUND in file client/pmix_client_get.c at line 334
b1303666 Minor cleanups to cherry-picks - PMIX ERROR: NOT-FOUND in file client/pmix_client_get.c at line 334
4f67b644 Cleanup handling of data requests for different nspaces - compile error
fddd7e98 Fix legacy support for PMI-1 - compile error
202b9bb4 (tag: v3.1.4rc2) Update VERSION to rc2 - passes
```

**Comment by jjhursey** [comment]
Any update on this issue?

**Comment by jjhursey** [comment]
Per March 25, 2021 teleconf
 * General feeling that this has been fixed in the master and v4.x releases. May also be fixed in v3.2
 * Please reopen if the problem persists.
